src/plugins/compression.cpp

```diff
@@ -0,0 +1,91 @@
+#include "compression.h"
+
+std::unique_ptr<Compression> Compression::create(std::string encoding) {
+#ifdef ENABLE_LZ4
+    if (encoding.compare("LZ4") == 0) {
+        return std::make_unique<Lz4Compression>();
+    }
+    else
+#endif /* ENABLE_LZ4 */
+    if (encoding.compare("None") == 0) {
+        return std::make_unique<NoCompression>();
+    }
+    else {
+        return std::unique_ptr<Compression>(nullptr);
+    }
+}
+
+#ifdef ENABLE_LZ4
+int Lz4Compression::compress(ByteIterator uncompressedStart, ByteIterator uncompressedEnd, std::vector<char>& compressed) {
+    uncompressedSize = uncompressedEnd-uncompressedStart;
+    compressed.resize(uncompressedSize+sizeof(size_t));
+    // compress the uncompressed source
+    auto actualCompressedSize = LZ4_compress_default(
+            &(*uncompressedStart),
+            compressed.data()+sizeof(size_t),
+            uncompressedSize,
+            uncompressedSize);
+    if (actualCompressedSize < 0) {
+        error = CompressionError::COMP_ERR_COMPRESSION_FAILED;
+        return static_cast<int>(actualCompressedSize);
```

Comment:
should this return `error` instead?

Reply:
For other backends (e.g. when I add zip, or lossy bit truncation or whatever), this return value can be used to give backend-specific error codes (think of it as developer errors).
An end user should probably only check getError(), which either returns COMP_SUCCESS or not.